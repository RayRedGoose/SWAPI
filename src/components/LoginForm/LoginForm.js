import React, { Component } from 'react';
import './LoginForm.scss'

export default class LoginForm extends Component {
  constructor() {
    super()
    this.state = {}
  }

  render() {
    return (
      <form className="loginForm">
        <h1 className="loginHeader">SWAPI Trivia</h1>
        <section className="nameInputAndLabel">
          <label></label>
          <input></input>
        </section>
        <section className="quoteInputAndLabel">
          <label></label>
          <input></input>
        </section>
        <section className="triviaLevelButtons">
          <input type="radio">Novice</input>
          <input type="radio">Intermediate</input>
          <input type="radio">Expert</input>
        </section>
        <button>Submit</button>
      </form>
    )
  }
}
