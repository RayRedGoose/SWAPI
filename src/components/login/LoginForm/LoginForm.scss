$font-dark: #5A2C05;
$border-color: #F1AA61;

.login-form {
  width: 40%;
  background: rgba(255, 255, 255, 0.9);
  margin: auto;
  padding: 20px;
  color: $font-dark;
  text-align: center;
  h1 {
    font-size: 72px;
    margin: 20px 0;
  }
  button {
    margin: 40px 0;
    font-size: 18px;
    padding: 0.4em 2em;
    background-color: $border-color;
    border: 2px solid $font-dark;
    text-transform: uppercase;
  }
}

.login-radio-buttons {
  padding: 20px 0;
  margin: auto;
  width: 60%;
  text-align: center;
  justify-content: space-between;
  display: flex;
}

.login {
  text-decoration: none;
  color: #5A2C05;
}
