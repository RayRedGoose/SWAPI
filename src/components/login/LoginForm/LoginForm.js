import React, { Component } from 'react'
import { NavLink } from 'react-router-dom';
import './LoginForm.scss'
import LoginInput from '../LoginInput/LoginInput'
import LoginRadioButton from '../LoginRadioButton/LoginRadioButton'
import ErrorNotification from '../ErrorNotification/ErrorNotification'

export default class LoginForm extends Component {
  constructor(props) {
    super()
    this.state = {
      data: {
        name: '',
        quote: '',
        rank: 'novice'
      },
      error: false
    }
  }

  toggleChangeState = (type, value) => {
    switch (type) {
      case 'name':
        this.setState({data: {...this.state.data, name: value}, error: false})
        break;
      case 'favorite quote':
        this.setState({data: {...this.state.data, quote: value}, error: false})
        break;
      default:
        this.setState({data: {...this.state.data, rank: value}, error: false})
        break;
    }
  }

  submitUser = (event) => {
    event.preventDefault()
    if (!this.checkError()) {
      this.props.addUser({...this.state.data})
      this.clearInputs()
    } else {
      this.setState({...this.state, error: true})
    }
  }

  checkError = () => {
    const valName = this.state.data.name.replace(' ', '') === ''
    const valQuote = this.state.data.quote.replace(' ', '') === ''
    const result = (valName || valQuote) ? true : false
    return result
  }

  clearInputs = () => {
    this.setState({data: {name: '', quote: '', rank: 'novice'}, error: false})
  }

  render() {
    const error = (this.state.error) ? <ErrorNotification /> : null

    const inputs = ['name', 'quote'].map(key => {
      const type = (key !== 'name') ? 'favorite quote' : key
      return (
        <LoginInput
          key={key}
          type={type}
          value={this.state.data[key]}
          toggleChangeState={this.toggleChangeState}/>
    )})

    const radios = ['novice', 'intermediate', 'expert'].map(rank => (
      <LoginRadioButton
        key={rank}
        type={rank}
        currentRank={this.state.data.rank}
        toggleChangeState={this.toggleChangeState}/>
    ))

    return (
      <form className="login-form">
        <h1>SWAPI Trivia</h1>
          {error}
          {inputs}
        <section className="login-radio-buttons">
          {radios}
        </section>
        <button onClick={this.submitUser}>
          <NavLink to='/movies' className="login">Submit</NavLink>
        </button>
      </form>
    )
  }
}
