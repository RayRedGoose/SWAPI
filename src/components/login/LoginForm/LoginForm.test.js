import React from 'react'
import {shallow} from 'enzyme'
import LoginForm from './LoginForm'

describe('LoginForm', () => {
  let form
  const addUser = jest.fn()

  beforeEach(() => {
    form = shallow(<LoginForm addUser={addUser} />)
  })

  it ('should match the snapshot with all data passed in correctly', () => {
    expect(form).toMatchSnapshot()
  })

  it ('should match the snapshot if toMoviesPage state is true', () => {
    form.setState({toMoviesPage: true})
    
    expect(form).toMatchSnapshot()
  })

  it ('should match the snapshot with error notification', () => {
    form.setState({error: true})

    expect(form).toMatchSnapshot()
  })

  describe("toggleChangeState", () => {

    it('should update name state when toggleChangeState is called with first argument as a name', () => {

      expect(form.state('data').name).toEqual('')

      form.instance().toggleChangeState('name', 'Yoda')

      expect(form.state('data').name).toEqual('Yoda')
    })

    it('should update quote state when toggleChangeState is called with first argument as a favorite quote', () => {

      expect(form.state('data').quote).toEqual('')

      form.instance().toggleChangeState('favorite quote', 'Do. Or do not. There is no try.')

      expect(form.state('data').quote).toEqual('Do. Or do not. There is no try.')
    })

    it('should update rank state when toggleChangeState is called with first argument as a rank', () => {

      expect(form.state('data').rank).toEqual('novice')

      form.instance().toggleChangeState('rank', 'expert')

      expect(form.state('data').rank).toEqual('expert')
    })
  })

  describe("submitUser", () => {

    it('should call submitUser when button was clicked', () => {
      const spy = jest.spyOn(form.instance(), 'submitUser').mockImplementation(() => {})
      form.instance().forceUpdate()

      form.find('button').simulate(
        'click',
        {preventDefault: jest.fn()}
      )

      expect(spy).toHaveBeenCalled()
    })

    it('should call checkError when submitUser is called', () => {
      const spy = jest.spyOn(form.instance(), 'checkError').mockImplementation(() => {})
      form.instance().forceUpdate()

      form.instance().submitUser(
        {preventDefault: jest.fn()}
      )

      expect(spy).toHaveBeenCalled()
    })

    it('should set error to state when button was clicked but all inputs are not filled out', () => {

      expect(form.state('error')).toEqual(false)

      form.find('button').simulate(
        'click',
        {preventDefault: jest.fn()}
      )

      expect(form.state('error')).toEqual(true)

    })

    describe("redirect", () => {

      it('should call redirect method if submitUser is called and all inputs are filled out', () => {
        const spy = jest.spyOn(form.instance(), 'redirect').mockImplementation(() => {})
        form.instance().forceUpdate()

        form.setState({data: {
          name: 'Yoda',
          quote: 'Do. Or do not. There is no try.'
        }})

        form.instance().submitUser(
          {preventDefault: jest.fn()}
        )

        expect(spy).toHaveBeenCalled()
      })

      it('should call addUser prop if redirect is called', () => {
        form.instance().redirect()

        expect(addUser).toHaveBeenCalled()
      })

      it('should call clearInputs method if redirect is called', () => {
        const spy = jest.spyOn(form.instance(), 'clearInputs').mockImplementation(() => {})
        form.instance().forceUpdate()

        form.instance().redirect()

        expect(spy).toHaveBeenCalled()
      })

      it('should change toMoviesPage state if redirect is called', () => {

        expect(form.state('toMoviesPage')).toEqual(false)

        form.instance().redirect()

        expect(form.state('toMoviesPage')).toEqual(true)
      })
    })
  })
})
