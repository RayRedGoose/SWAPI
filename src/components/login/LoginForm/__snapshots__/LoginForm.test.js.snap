// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm should match the snapshot with all data passed in correctly 1`] = `
<form
  className="login-form"
>
  <h1>
    SWAPI Trivia
  </h1>
  <LoginInput
    key="name"
    toggleChangeState={[Function]}
    type="name"
    value=""
  />
  <LoginInput
    key="quote"
    toggleChangeState={[Function]}
    type="favorite quote"
    value=""
  />
  <section
    className="login-radio-buttons"
  >
    <LoginRadioButton
      currentRank="novice"
      key="novice"
      toggleChangeState={[Function]}
      type="novice"
    />
    <LoginRadioButton
      currentRank="novice"
      key="intermediate"
      toggleChangeState={[Function]}
      type="intermediate"
    />
    <LoginRadioButton
      currentRank="novice"
      key="expert"
      toggleChangeState={[Function]}
      type="expert"
    />
  </section>
  <button
    onClick={[Function]}
  >
    Submit
  </button>
</form>
`;

exports[`LoginForm should match the snapshot with error notification 1`] = `
<form
  className="login-form"
>
  <h1>
    SWAPI Trivia
  </h1>
  <ErrorNotification />
  <LoginInput
    key="name"
    toggleChangeState={[Function]}
    type="name"
    value=""
  />
  <LoginInput
    key="quote"
    toggleChangeState={[Function]}
    type="favorite quote"
    value=""
  />
  <section
    className="login-radio-buttons"
  >
    <LoginRadioButton
      currentRank="novice"
      key="novice"
      toggleChangeState={[Function]}
      type="novice"
    />
    <LoginRadioButton
      currentRank="novice"
      key="intermediate"
      toggleChangeState={[Function]}
      type="intermediate"
    />
    <LoginRadioButton
      currentRank="novice"
      key="expert"
      toggleChangeState={[Function]}
      type="expert"
    />
  </section>
  <button
    onClick={[Function]}
  >
    Submit
  </button>
</form>
`;
